packages/ember-application/lib/utils/validate-type.js

```diff
@@ -0,0 +1,24 @@
+/**
+@module ember
+@submodule ember-application
+*/
+
+import Route from "ember-routing/system/route";
+import Component from "ember-views/views/component";
+import View from "ember-views/views/view";
+import Service from "ember-runtime/system/service";
+
+let VALIDATED_TYPES = {
+  route: Route,
+  component: Component,
+  view: View,
+  service: Service
+};
+
+export default function validateType(resolvedType, parsedName) {
+  let expectedType = VALIDATED_TYPES[parsedName.type];
+
+  Ember.assert(`Expected ${parsedName.fullName} to resolve to an ${expectedType} but instead it was ${resolvedType}.`, function() {
+    return !expectedType || expectedType.detect(resolvedType);
```

Comment:
I'm not understanding the object stamping approach yet. It seems like if components can be POJOs then they shouldn't be validated. This kind of validation only makes sense to me for objects that are integrated with Ember through private APIs.